mlflow/store/artifact/s3_artifact_repo.py

```diff
@@ -150,44 +179,135 @@ def _upload_file(self, s3_client, local_file, bucket, key):
         s3_client.upload_file(Filename=local_file, Bucket=bucket, Key=key, ExtraArgs=extra_args)
 
     def log_artifact(self, local_file, artifact_path=None):
-        (bucket, dest_path) = self.parse_s3_compliant_uri(self.artifact_uri)
+        dest_path = self.bucket_path
         if artifact_path:
             dest_path = posixpath.join(dest_path, artifact_path)
         dest_path = posixpath.join(dest_path, os.path.basename(local_file))
+        key = os.path.normpath(dest_path)
         self._upload_file(
-            s3_client=self._get_s3_client(), local_file=local_file, bucket=bucket, key=dest_path
+            s3_client=self._get_s3_client(),
+            local_file=local_file,
+            bucket=self.bucket,
+            key=key,
         )
 
-    def log_artifacts(self, local_dir, artifact_path=None):
-        (bucket, dest_path) = self.parse_s3_compliant_uri(self.artifact_uri)
-        if artifact_path:
-            dest_path = posixpath.join(dest_path, artifact_path)
-        s3_client = self._get_s3_client()
-        local_dir = os.path.abspath(local_dir)
-        for root, _, filenames in os.walk(local_dir):
-            upload_path = dest_path
-            if root != local_dir:
-                rel_path = os.path.relpath(root, local_dir)
-                rel_path = relative_path_to_artifact_path(rel_path)
-                upload_path = posixpath.join(dest_path, rel_path)
-            for f in filenames:
-                self._upload_file(
-                    s3_client=s3_client,
-                    local_file=os.path.join(root, f),
-                    bucket=bucket,
-                    key=posixpath.join(upload_path, f),
+    def _get_write_credential_infos(self, remote_file_paths):
+        """
+        Instead of returning ArtifactCredentialInfo objects, we instead return a list of initialized
+        S3 client. We do so because S3 clients cannot be instantiated within each thread.
+        """
+        return [self._get_s3_client() for _ in remote_file_paths]
+
+    def _upload_to_cloud(self, cloud_credential_info, src_file_path, artifact_file_path):
+        dest_path = posixpath.join(self.bucket_path, artifact_file_path)
+        key = os.path.normpath(dest_path)
+        if (
+            MLFLOW_ENABLE_MULTIPART_UPLOAD.get()
+            and os.path.getsize(src_file_path) > _MULTIPART_UPLOAD_CHUNK_SIZE
+        ):
+            self._multipart_upload(cloud_credential_info, src_file_path, self.bucket, key)
+        else:
+            self._upload_file(cloud_credential_info, src_file_path, self.bucket, key)
+
+    def _multipart_upload(self, cloud_credential_info, local_file, bucket, key):
```

Comment:
Do we need this? `boto3` automatically uses MPU when uploading a large file.

Reply:
Our custom implementation achieves a 3.5x speed up for uploads and downloads. The boto3 client is not well optimized.

Reply:
> The boto3 client is not well optimized.

I believe boto3 client is well-optimized. It's been battle-tested for so long. Can you upload a single large file and compare the performance?

Reply:
If you look at the PR description, downloading and uploading llama2 achieved 3.x speedups compared to master (which uses the vanilla boto3 API).

Reply:
But I can also test uploading and downloading a single large file tomorrow morning.

Reply:
How many large files does llama2 70b or 13b have? `master` uses the vanilla boto3 API without parallelization on files, but this doesn't necessarily mean boto3 is slow.

Reply:
That's a good point. I've kicked off two runs ([feature branch](https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#job/145048119242984/run/1099009648951780), [master](https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#job/145048119242984/run/297503975833865)) of uploading and downloading one 30 GB file. So far, it still seems like it is the case that the feature branch is 3.x times faster for both uploads and downloads.

Reply:
@harupy we had the same question, we were surprised to find that boto3 does indeed not perform well; separate from upload performance, we saw slowness with boto3 for downloads (as shown in Jerry's benchmark).

Reply:
@jerrylian-db one question though, could we reuse more logic from the `DatabricksArtifactRepository` here? I.e. the logic around submitting chunks for upload/download to a threadpool

Reply:
The end goal of this PR is to basically leverage the existing fast-upload and download logic written for `DatabricksArtifactRepository` for generic S3 artifact repo + models in UC codepaths

Reply:
@smurching the multipart implementation here actually has quite a few differences with the `DatabricksArtifactRepository` implementation. I think it would be more work than it is worth to try to refactor both to share a common code path and then verify the changes.

Reply:
@jerrylian-db @smurching Can you provide more info about the motivation here? Just to confirm, the use case is that a customer has a large model stored in an external S3 bucket that they're using with MLflow Tracking & they want to upload to UC Model Registry? If so, how common / urgent is this use case? Asking because we need to make a release in the next day or two & I'm wary of including significant last-minute changes to the S3 client widely used by OSS users for what seems like a small fraction of Databricks customers.

Reply:
Yeah I don't think this needs to be strictly in MLflow 2.7 - we just want it in an open source MLflow release by the week of 9/26 (for a marketing announcement). If we can make another OSS release before then I think that's fine!

Reply:
> the use case is that a customer has a large model stored in an external S3 bucket that they're using with MLflow Tracking & they want to upload to UC Model Registry

The use case is to accelerate model registration/download to/from UC. When a model is registered/loaded from UC, the client gets temporary credentials and calls into the S3ArtifactRepository to write/read models

Reply:
Got it - makes sense. 2.7.1 before 9/26 sounds great :)

Reply:
Agree 👍